### DISCUSSION TITLE
Renovate keeps hanging.

### DISCUSSION THREAD

**Discussion by LaravDuin** [renovatebot/renovate#16895] [Request Help]
In our organisation we are using a self-hosted instance of renovate bot to update around 700 to 800 repositories. Unfortunately the job keeps hanging frequently. We noticed that it is often after logs involving git tasks, therefore we were hoping that the issue would be fixed after configuring this gitTimeout. Unfortunately we keep seeing the same issue.

I'm wondering if I could get some assistance, am I missing some other config, or maybe there is a bug?

When I'm playing around on a local project with a very small gitTimeout, I do see the following output when the git task takes longer:
```
{..."hostType":"git","err":{"task":{"commands":["clone","--filter=blob:none","https://**redacted**@some-project.git","."],"format":"utf-8","parser":"[function]"},"plugin":"timeout","message":"block timeout reached","stack":"Error: block timeout reached\n    at Timeout.kill (/usr/local/lib/node_modules/renovate/node_modules/simple-git/dist/cjs/index.js:1328:24)\n    at listOnTimeout (internal/timers.js:557:17)\n    at processTimers (internal/timers.js:500:7)"},"msg":"Host error","time":"2022-08-01T10:44:33.794Z","v":0}
```

We use github-enterprise platform
renovate image: https://github.com/whitesource/renovate-on-prem/releases/tag/2.4.1

In the logs I see that gitTimeout is set as a value in the config:
```
{"name":"renovate","hostname":"worker018","pid":7,"level":20,"logContext":"_ghzksJ04KqA2odCei3K7","config":{"autodiscover":true,"autodiscoverFilter":"some-repo/*","dryRun":null,"onboarding":false,"onboardingConfig":{"extends":["config:base"]},"hostRules":[{"hostType":"github","matchHost":"github.com","token":"***********"}],"binarySource":"install","platform":"github","endpoint":"https://github.*/api/v3/","gitTimeout":60000},"msg":"Combined config","time":"2022-08-01T10:06:56.379Z","v":0}
```

Some last logs that projects seem to hang after:
```
____________________________________________________________________
"lastJobDispatchDate": "2022-07-30T13:46:39.265Z"
"repository":"some-project","msg":"Performing blobless clone","time":"2022-07-30T13:46:40.032Z"

____________________________________________________________________
"lastJobDispatchDate": "2022-08-01T08:58:30.254Z"
"repository":"some-project","msg":"Using app token for git init","time":"2022-08-01T08:58:30.700Z"

____________________________________________________________________
"lastJobDispatchDate": "2022-08-01T11:20:42.391Z"
"repository":"some-project","msg":"Performing blobless clone","time":"2022-08-01T11:20:43.036Z"
```

Please let me know if any other info is useful to share.

**Comment by PhilipAbed** [comment]
just a hunch, but have you tried limiting the number of repositories in each run? maybe there isn't enough memory in your container/image to support all 800 repos at once?

**Reply by LaravDuin** [reply]
I don't think it is related to the number of repositories. Sometimes the job hangs after one hour of restarting and another time is takes a full day.

**Comment by nabeelsaabna** [comment]
is it simply taking too much time to finish the `git` command ?
or stuck hanging for hours ?

**Reply by LaravDuin** [reply]
We have to restart the job in order to get it started again. We have configured our renovate instance to schedule all projects at XX:00 hours. The scheduling is happening in parallel to updating our projects. This process is happily continuing, however, updating a project, once it's stuck, it hangs indefinitely. 

Also, the last log indicates it is around a git task, but I can't say that I'm 100% sure. I can only indicate it hangs in between two debug log lines.

**Reply by PhilipAbed** [reply]
did you say you run in parallel? i do not believe that is a good idea, since renovate uses the same cache, both jobs could be overlapping causing this. or is it in a different container runs with different cache?

**Reply by LaravDuin** [reply]
It is all running in the same docker container.

Let me add some more details. This is what I'm seeing in the logs itself. 

I see logs of a project being updated, then during those logs when it's time to que new jobs (every XX:00 hours), I see also logs fetching all projects that need updating. Limit of fetching is 100, so I see 8 requests. I believe that this process is happening in parallel with updating the dependencies of the current project. 
When all projects, that require updating, are fetched, then the current project that is being updated will be paused and the queuing process will start. After that is finished the current project will continue updating.

We are not updating multiple projects in parallel, that param is set to `max_parallel: 1` in our `config.yaml` file.

**Reply by PhilipAbed** [reply]
let's do some tests:
1) try to run git clone on the repository from the logs, see how much time it really takes to clone. 

2) dont use auto-discover, take this repository that is timing out, put it in your config.js(global config) inside this flag: 
  `repositories: ['org/project'],` 
run renovate on it alone dont run on others for now let's see if it's failing.

another thing i'm wondering about, are you using proxy? could that be the actual problem?

**Reply by LaravDuin** [reply]
I've double checked with our infra team and we are not using any proxies. We are not configuring any renovate proxy settings and we are also not using a proxy outside of those renovate settings.

On the topic of testing the git clone time, I've to say that it is not one particular job that seems to hang. The projects that are stuck seem to be random. Looking into the existing log files I can give you some numbers. So, yesterday, the following project seemed to hang on the following log line:

```
"repository":"gdpr-kafka-producer-lib","msg":"Performing blobless clone","time":"2022-08-03T09:06:25.218Z"
```

These are logs from the same project when it was updated just fine:
```
{"name":"renovate","hostname":"worker014","pid":7,"level":20,"logContext":"GiX90iSmW","repository":"gdpr-kafka-producer-lib","msg":"Performing blobless clone","time":"2022-08-04T06:33:17.491Z","v":0}
{"name":"renovate","hostname":"worker014","pid":7,"level":20,"logContext":"GiX90iSmW","repository":"gdpr-kafka-producer-lib","durationMs":525,"msg":"git clone completed","time":"2022-08-04T06:33:18.016Z","v":0}
```

So, that takes 525 ms to complete. 


Another thing I did not mention before, but is important to understand the whole situation: The queuing job, which is scheduled every hour, is not being blocked when a project is hanging.

**Reply by nabeelsaabna** [reply]
might be related to `fs` rather than `git` 
https://github.com/renovatebot/renovate/blob/1bc8fee4947d65acd18113069b977cc91a9be4cc/lib/util/git/index.ts#L388-L391 

`localDir` is printed at the start of the function and not being used until `fs.emptyDir(localDir);`
https://github.com/renovatebot/renovate/blob/1bc8fee4947d65acd18113069b977cc91a9be4cc/lib/util/git/index.ts#L346

**Reply by LaravDuin** [reply]
Could be for the `Performing blobless clone` logs. 

However, a second point where it often gets stuck is:
```
{"name":"renovate","hostname":"worker014","pid":7,"level":20,"logContext":"GiX90iSmW","repository":"service-pipeline-library","msg":"Using app token for git init","time":"2022-08-01T08:58:30.700Z","v":0}
```

When it's successful the next log line would have been:
```
{"name":"renovate","hostname":"worker014","pid":7,"level":20,"logContext":"GiX90iSmW","repository":"service-pipeline-library","msg":"Using app token for git init","time":"2022-08-04T06:32:02.075Z","v":0}
{"name":"renovate","hostname":"worker014","pid":7,"level":20,"logContext":"GiX90iSmW","repository":"service-pipeline-library","msg":"Resetting npmrc","time":"2022-08-04T06:32:02.532Z","v":0}
``` 

So far these are the only two log lines where the process seems to have to possibility of hanging.
